File: dart/test/variadic_user_function_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:test/test.dart';

Map<String, Object?> _contract() {
  return (jsonDecode(
            File(
              '../capability_conformance/callable_signature_contract.json',
            ).readAsStringSync(),
          )
          as Map)
      .cast<String, Object?>();
}

String _fixtureSource() {
  final fixture = (_contract()['fixture']! as Map).cast<String, Object?>();
  return fixture['spec_source']! as String;
}

CompiledSpec _compileSource(String source) {
  final parsed = parseSpecWithStagedUserFunctionDefinitions(source);
  validateSpec(parsed);
  return compileSpec(parsed);
}

Object? _expectedFixtureValue() {
  final fixture = (_contract()['fixture']! as Map).cast<String, Object?>();
  return fixture['expected'];
}

void main() {
  test('definition shell emits exact v1 and v2 signature unions', () {
    final nodes = parseUserFunctionDefinitionAsts(_fixtureSource());
    expect(nodes, hasLength(3));

    final fixed = (nodes[0] as Map).cast<String, Object?>();
    expect(fixed['version'], 1);
    expect(fixed['params'], ['left', 'right']);
    expect(fixed['arity'], 2);
    expect(fixed, isNot(contains('signature')));

    for (final item in [
      (node: nodes[1], name: 'all_values', positional: <String>[], minimum: 0),
      (node: nodes[2], name: 'collect', positional: ['prefix'], minimum: 1),
    ]) {
      final node = (item.node as Map).cast<String, Object?>();
      final signature = (node['signature']! as Map).cast<String, Object?>();
      expect(node['version'], 2, reason: item.name);
      expect(node['name'], item.name);
      expect(node, isNot(contains('params')));
      expect(node, isNot(contains('arity')));
      expect(signature, {
        'kind': 'callable_signature',
        'version': 1,
        'positional_params': item.positional,
        'rest_param': 'items',
        'min_arity': item.minimum,
        'max_arity': null,
      });
      for (final staged in ['body_payload', 'body_parse_job']) {
        final record = (node[staged]! as Map).cast<String, Object?>();
        expect(record['signature'], signature);
        expect(record, isNot(contains('params')));
        expect(record, isNot(contains('arity')));
      }
    }
  });

  test('descriptor uses exact v2 keys and native fixture result', () {
    final compiled = _compileSource(_fixtureSource());
    final descriptor = compiled.descriptorState.toJson();
    final contract = _contract();
    final versions = (contract['definition_versions']! as Map)
        .cast<String, Object?>();
    final variadic = (versions['variadic']! as Map).cast<String, Object?>();
    final expectedKeys = (variadic['record_fields']! as List).cast<String>();

    final functions = (descriptor['functions']! as Map).cast<String, Object?>();
    for (final name in ['all_values', 'collect']) {
      final record = (functions[name]! as Map).cast<String, Object?>();
      expect(record.keys.toSet(), expectedKeys.toSet());
      expect(record['version'], 2);
      expect(record, isNot(contains('params')));
      expect(record, isNot(contains('arity')));
    }

    final result = LinkedSpecRuntimeEngine(compiled).parse('xx');
    expect(result.value, _expectedFixtureValue());
  });

  test('evaluates once left-to-right and allocates fresh rest arrays', () {
    final compiled = _compileSource(r'''fn gather(...items) { return(items) }
fn mutate_rest(...items) { items += "mutated"; return(items) }

Top::
 /x/ -> Done {
   order = "";
   first = gather(order = cat(order, "a"), order = cat(order, "b"), order);
   return({
     "first" : first,
     "fresh_one" : mutate_rest("a"),
     "fresh_two" : mutate_rest(),
     "order" : order
   })
 }

Done::
 /x/
''');
    expect(LinkedSpecRuntimeEngine(compiled).parse('xx').value, {
      'first': ['a', 'ab', 'ab'],
      'fresh_one': ['a', 'mutated'],
      'fresh_two': ['mutated'],
      'order': 'ab',
    });
  });

  test('fixed and variadic arity failures stay distinct', () {
    final fixed = _fixtureSource().replaceFirst(
      'pair("left", "right")',
      'pair("left", "right", "extra")',
    );
    expect(
      () => LinkedSpecRuntimeEngine(_compileSource(fixed)).parse('xx'),
      throwsA(
        isA<RuntimeInterpreterException>().having(
          (error) => error.message,
          'message',
          contains("user function 'pair' expects 2 argument(s), got 3"),
        ),
      ),
    );

    final variadic = _fixtureSource().replaceFirst('collect("p")', 'collect()');
    expect(
      () => LinkedSpecRuntimeEngine(_compileSource(variadic)).parse('xx'),
      throwsA(
        isA<RuntimeInterpreterException>().having(
          (error) => error.message,
          'message',
          contains(
            "user function 'collect' expects at least 1 argument(s), got 0",
          ),
        ),
      ),
    );
  });

  test('rejects malformed rest definitions and keyword calls', () {
    for (final params in [
      '...items, tail',
      '...left, ...right',
      '...',
      'items...',
      '... items',
    ]) {
      final source = 'fn bad($params) { return(undef) }\nTop::\n /x/\n';
      expect(
        () => parseSpecWithStagedUserFunctionDefinitions(source),
        throwsA(anything),
        reason: params,
      );
    }

    for (final params in ['item, ...item', '...return']) {
      final source = 'fn bad($params) { return(undef) }\nTop::\n /x/\n';
      expect(
        () => validateSpec(parseSpecWithStagedUserFunctionDefinitions(source)),
        throwsA(isA<SpecValidationException>()),
        reason: params,
      );
    }

    final compiled = _compileSource(_fixtureSource());
    final expression = ActionCallExpr(
      source: 'collect(prefix = "p")',
      sourceSpan: const ActionSourceSpan(start: 0, end: 21),
      name: 'collect',
      args: [
        ActionKeywordArgument(
          name: 'prefix',
          value: ActionStringLiteralExpr(
            source: '"p"',
            sourceSpan: const ActionSourceSpan(start: 17, end: 20),
            value: 'p',
            quote: '"',
          ),
        ),
      ],
    );
    final resolution = resolveActionExpressionContracts(
      expression,
      functionRegistry: compiled.functionRegistry,
    );
    expect(
      resolution.diagnostics.single.code,
      'user_function_keyword_arguments_unsupported',
    );
  });

  test('generated state preserves and executes variadic signatures', () {
    final compiled = _compileSource(_fixtureSource());
    final plan = buildGeneratedRulePlan(compiled);
    expect(
      executeGeneratedParserV2(compiled, plan, 'xx', 'variadic-contract.spec'),
      _expectedFixtureValue(),
    );

    final generated = emitDartSourceV2(compiled, 'variadic-contract.spec');
    final encoded = RegExp(
      "const _compiledSpecJsonBase64 = '([^']+)';",
    ).firstMatch(generated)![1]!;
    final normalized = (jsonDecode(utf8.decode(base64Decode(encoded))) as Map)
        .cast<String, Object?>();
    final functions = (normalized['functions']! as List).cast<Object?>();
    final allValues = (functions[1]! as Map).cast<String, Object?>();
    expect(allValues, isNot(contains('params')));
    expect(allValues, isNot(contains('arity')));
    expect((allValues['signature']! as Map)['rest_param'], 'items');

    final roundTrip = compileSpec(SpecFile.fromJson(normalized));
    expect(
      LinkedSpecRuntimeEngine(roundTrip).parse('xx').value,
      _expectedFixtureValue(),
    );
  });
}

File: julia/test/variadic_user_function_contract_test.jl
const CALLABLE_SIGNATURE_CONTRACT = JSON3.read(
    read(joinpath(REPO_ROOT, "capability_conformance", "callable_signature_contract.json"), String),
    Dict{String,Any},
)

function _compile_variadic_source(source::AbstractString)
    spec = parse_spec_with_staged_user_function_definitions(source)
    validate_spec(spec)
    return compile_spec(spec)
end

function _callable_fixture_source()
    return CALLABLE_SIGNATURE_CONTRACT["fixture"]["spec_source"]
end

@testset "Julia variadic user-function contract" begin
    @testset "definition shell emits exact v1 and v2 signature unions" begin
        nodes = parse_user_function_definition_asts(_callable_fixture_source())
        @test length(nodes) == 3

        fixed = nodes[1]
        @test fixed["version"] == 1
        @test fixed["params"] == Any["left", "right"]
        @test fixed["arity"] == 2
        @test !haskey(fixed, "signature")

        for (node, name, positional, minimum) in [
            (nodes[2], "all_values", Any[], 0),
            (nodes[3], "collect", Any["prefix"], 1),
        ]
            signature = node["signature"]
            @test node["version"] == 2
            @test node["name"] == name
            @test !haskey(node, "params")
            @test !haskey(node, "arity")
            @test signature == Dict{String,Any}(
                "kind" => "callable_signature",
                "version" => 1,
                "positional_params" => positional,
                "rest_param" => "items",
                "min_arity" => minimum,
                "max_arity" => nothing,
            )
            for staged_name in ("body_payload", "body_parse_job")
                staged = node[staged_name]
                @test staged["signature"] == signature
                @test !haskey(staged, "params")
                @test !haskey(staged, "arity")
            end
        end
    end

    @testset "descriptor and native runtime honor the neutral fixture" begin
        compiled = _compile_variadic_source(_callable_fixture_source())
        descriptor = to_descriptor_json(compiled)
        expected_keys = Set(String.(
            CALLABLE_SIGNATURE_CONTRACT["definition_versions"]["variadic"]["record_fields"],
        ))
        for name in ("all_values", "collect")
            record = descriptor["functions"][name]
            @test Set(keys(record)) == expected_keys
            @test record["version"] == 2
            @test !haskey(record, "params")
            @test !haskey(record, "arity")
        end

        result = runtime_execute(LinkedSpecRuntimeEngine(compiled), "xx")
        @test result.value == CALLABLE_SIGNATURE_CONTRACT["fixture"]["expected"]
    end

    @testset "arguments evaluate once left-to-right and rest arrays are fresh" begin
        compiled = _compile_variadic_source(raw"""fn gather(...items) { return(items) }
fn mutate_rest(...items) { items += "mutated"; return(items) }

Top::
 /x/ -> Done {
   order = "";
   first = gather(order = cat(order, "a"), order = cat(order, "b"), order);
   return({
     "first" : first,
     "fresh_one" : mutate_rest("a"),
     "fresh_two" : mutate_rest(),
     "order" : order
   })
 }

Done::
 /x/
""")
        @test runtime_execute(LinkedSpecRuntimeEngine(compiled), "xx").value == Dict{String,Any}(
            "first" => Any["a", "ab", "ab"],
            "fresh_one" => Any["a", "mutated"],
            "fresh_two" => Any["mutated"],
            "order" => "ab",
        )
    end

    @testset "arity, malformed definitions, and keyword calls fail distinctly" begin
        fixed = replace(
            _callable_fixture_source(),
            "pair(\"left\", \"right\")" => "pair(\"left\", \"right\", \"extra\")";
            count = 1,
        )
        fixed_engine = LinkedSpecRuntimeEngine(_compile_variadic_source(fixed))
        @test_throws RuntimeInterpreterException runtime_execute(fixed_engine, "xx")
        try
            runtime_execute(fixed_engine, "xx")
        catch error
            @test occursin("user function 'pair' expects 2 argument(s), got 3", sprint(showerror, error))
        end

        variadic = replace(
            _callable_fixture_source(),
            "collect(\"p\")" => "collect()";
            count = 1,
        )
        variadic_engine = LinkedSpecRuntimeEngine(_compile_variadic_source(variadic))
        @test_throws RuntimeInterpreterException runtime_execute(variadic_engine, "xx")
        try
            runtime_execute(variadic_engine, "xx")
        catch error
            @test occursin("user function 'collect' expects at least 1 argument(s), got 0", sprint(showerror, error))
        end

        for params in ("...items, tail", "...left, ...right", "...", "items...", "... items")
            source = "fn bad($params) { return(undef) }\nTop::\n /x/\n"
            @test_throws Exception parse_spec_with_staged_user_function_definitions(source)
        end
        for params in ("item, ...item", "...return")
            source = "fn bad($params) { return(undef) }\nTop::\n /x/\n"
            @test_throws SpecValidationException validate_spec(
                parse_spec_with_staged_user_function_definitions(source),
            )
        end

        compiled = _compile_variadic_source(_callable_fixture_source())
        expression = ActionCallExpr(
            source = "collect(prefix = \"p\")",
            source_span = ActionSourceSpan(0, 21),
            name = "collect",
            args = [ActionKeywordArgument(
                name = "prefix",
                value = ActionStringLiteralExpr(
                    source = "\"p\"",
                    source_span = ActionSourceSpan(17, 20),
                    value = "p",
                    quote_char = "\"",
                ),
            )],
        )
        resolution = resolve_action_expression_contracts(
            expression;
            function_registry = compiled.function_registry,
        )
        @test only(resolution.diagnostics).code == "user_function_keyword_arguments_unsupported"
    end

    @testset "generated state preserves and executes variadic signatures" begin
        compiled = _compile_variadic_source(_callable_fixture_source())
        plan = build_generated_rule_plan(compiled)
        @test execute_generated_parser_v2(
            compiled,
            plan,
            "xx",
            "variadic-contract.spec",
        ) == CALLABLE_SIGNATURE_CONTRACT["fixture"]["expected"]

        generated = emit_julia_source_v2(compiled, "variadic-contract.spec")
        matched = match(r"const _COMPILED_SPEC_JSON_HEX = \"([^\"]+)\"", generated)
        @test matched !== nothing
        normalized = JSON3.read(String(hex2bytes(matched.captures[1])), Dict{String,Any})
        all_values = normalized["functions"][2]
        @test !haskey(all_values, "params")
        @test !haskey(all_values, "arity")
        @test all_values["signature"]["rest_param"] == "items"

        round_trip = compile_spec(from_json(SpecFile, normalized))
        @test runtime_execute(LinkedSpecRuntimeEngine(round_trip), "xx").value ==
              CALLABLE_SIGNATURE_CONTRACT["fixture"]["expected"]
    end
end
